chevah/compat/testing/testcase.py

```diff
@@ -807,18 +807,19 @@ def _get_os_version():
 
     On Linux is the distribution name, without the version.
 
-    On Windows it is the `nt` followed by the major and minor NT version.
-    IS not the marketing name.
-    We only support Windows NT family.
+    On Windows it is the `nt` followed by the version, which
+    consists of major version, minor version, build number.
+    It is not the marketing name.
+    We only support the Windows NT family.
     See: https://en.wikipedia.org/wiki/Windows_NT#Releases
 
     On OSX it returns `osx` followed by the version.
-    Is not the Darwin version.
+    It is not the version of the underlying Darwin OS.
     See: https://en.wikipedia.org/wiki/MacOS#Release_history
     """
     if os.name == 'nt':
-        parts = platform.version().split('.')
-        return 'nt-%s.%s' % (parts[0], parts[1])
+        parts = platform.win32_ver()[1].split('.')
+        return 'nt-%s.%s.%s' % (parts[0], parts[1], parts[2])
```

Comment:
This needs to be reverted, and have a separate `os_build_number` or just `os_build` member.

we have os_family, os_name, os_version which are all not comparable.
They were created to make tests easy to write and read and to document the code itself, without the need for extra comments.

The current usage as `int(self.os_version.split('.')[2]) >= 14393` is not easy to write or read.

What is `14393` ... if you always have to go to https://en.wikipedia.org/wiki/Windows_10_version_history and see that 14393 is in fact Windows 10 1607

I would lilke to see something like

`if self.os_build >= self.os_builds.WIN10_1607`

or


`if self.os_build >= self.WIN10_1607`


---------

The build is comparable, but as an int... so each time we will have to use the build number in a test, we will have to extract it from the os_version, convert in int and then compare.

Is better to have the os_build available right away and a few constants are reference points.

Reply:
Thanks for the review! 

I agree this OS version with a build number is an ugly problem. And probably could be addressed more elegantly. 

But I feel you haven't addressed the core issue at hand here. And that is having identical `os_version` strings for different versions of Windows. For now, only Windows Server 2016 and 2019 are affected, but it's possible that in the near future all supported version of Windows will be in the same situation!

The only argument I heard against this is _maybe they are not different and we should treat them the same_. But the need for separate changes here show they are not the same. Microsoft doesn't think they are the same. We also test them separately, with dedicated slaves etc. This problem will creep up in more and more places.

So my stance is that we have to improve `os_version` to *always* differentiate between major OS versions. As shown in my meagre attempt, the required changes are not that invasive.

**P.S.** Now that I look at the code with a fresh eye, I can see how we got here. Because for Windows we chose to only look at the first two "parts" of `platform.version()`. This was based on an assumption that proved wrong. I think we should not try to be smarter than upstream Python here, therefore we should use the whole version string. That includes the build number for a reason.